csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/InstructionTag.qll

```diff
@@ -0,0 +1,164 @@
+import csharp
+import semmle.code.csharp.ir.Util
+
+private predicate elementIsInitialized(int elementIndex) {
+  exists(ArrayInitWithMod initList |
+    initList.isInitialized(elementIndex)
+  )
+}
+
+newtype TInstructionTag =
+  OnlyInstructionTag() or  // Single instruction (not including implicit Load)
+  InitializeThisTag() or
+  InitializerVariableAddressTag() or
+  InitializerLoadStringTag() or
+  InitializerStoreTag() or
+  ZeroPadStringConstantTag() or
+  ZeroPadStringElementIndexTag() or
+  ZeroPadStringElementAddressTag() or
+  ZeroPadStringStoreTag() or
+  AssignOperationLoadTag() or
+  AssignOperationConvertLeftTag() or
+  AssignOperationOpTag() or
+  AssignmentConvertRightTag() or
+  AssignOperationConvertResultTag() or
+  AssignmentStoreTag() or
+  CrementLoadTag() or
+  CrementConstantTag() or
+  CrementOpTag() or
+  CrementStoreTag() or
+  EnterFunctionTag() or
+  ReturnValueAddressTag() or
+  ReturnTag() or
+  ExitFunctionTag() or
+  UnmodeledDefinitionTag() or
+  UnmodeledUseTag() or
+  AliasedDefinitionTag() or
+  SwitchBranchTag() or
+  CallTargetTag() or
+  CallTag() or
+  CallSideEffectTag() or
+  AllocationSizeTag() or
+  AllocationElementSizeTag() or
+  AllocationExtentConvertTag() or
+  ValueConditionConditionalBranchTag() or
+  ConditionValueTrueTempAddressTag() or
+  ConditionValueTrueConstantTag() or
+  ConditionValueTrueStoreTag() or
+  ConditionValueFalseTempAddressTag() or
+  ConditionValueFalseConstantTag() or
+  ConditionValueFalseStoreTag() or
+  ConditionValueResultTempAddressTag() or
+  ConditionValueResultLoadTag() or
+  BoolConversionConstantTag() or
+  BoolConversionCompareTag() or
+  LoadTag() or  // Implicit load due to lvalue-to-rvalue conversion
+  CatchTag() or
+  ThrowTag() or
+  UnwindTag() or
+  InitializerUninitializedTag() or
+  InitializerElementIndexTag(int elementIndex) {
+    elementIsInitialized(elementIndex)
+  } or
+  InitializerElementAddressTag(int elementIndex) {
+    elementIsInitialized(elementIndex)
+  } or
+  InitializerElementDefaultValueTag(int elementIndex) {
+    elementIsInitialized(elementIndex)
+  } or
+  InitializerElementDefaultValueStoreTag(int elementIndex) {
+    elementIsInitialized(elementIndex)
+  } or
+  // Added for C#  
+  NewObjTag() or
+  // TODO: remove the need for indexing
+  PointerAddTag(int index) {
```

Comment:
I would like to refactor that so that each index creates a new `TranslatedExpression`; this way we only use `getChild` and `getChildSuccessor`, removing the need for having indexed `PointerAddTag`s and `ElementsAddressTag`s.